subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultBuildLogicBuildQueue.java

```diff
@@ -73,20 +87,20 @@ private <T> T doBuild(List<TaskIdentifier.TaskBasedTaskIdentifier> tasks, Suppli
         return continuationUnderLock.get();
     }
 
-    @SuppressWarnings("try")
     private <T> T withBuildLogicQueueLock(Supplier<T> buildAction) {
-        lock.lock();
-        try {
-            final boolean firstLockHolder = lock.getHoldCount() == 1;
-            if (firstLockHolder) { // lock file at the top of the callstack only
-                try (FileLock ignored = lockBuildLogicQueueFile()) {
-                    return buildAction.get();
+        return resource.withLock(() -> {
```

Comment:
🤔 Is `Synchronizer` usage a game changer in this PR? Otherwise, I can't understand why this code should provide the behavior mentioned in the PR's title

Reply:
> Is `Synchronizer` usage a game changer in this PR?

Yes as it releases the worker lease before waiting.